cdap-tms/src/main/java/co/cask/cdap/messaging/store/cache/CachingMessageTable.java

```diff
@@ -0,0 +1,386 @@
+  }
+
+  @Override
+  public void store(Iterator<? extends Entry> entries) throws IOException {
+    // Write it to the message table first
+    CopyingIterator iterator = new CopyingIterator(entries);
+    messageTable.store(iterator);
+
+    Multimap<TopicId, Entry> topicEntries = iterator.getEntries();
+    for (Map.Entry<TopicId, Collection<Entry>> entry : topicEntries.asMap().entrySet()) {
+      MessageCache<Entry> messageCache = cacheProvider.getMessageCache(entry.getKey());
+      // Write it to the cache if it is enabled for the topic
+      if (messageCache != null) {
+        messageCache.addAll(entry.getValue().iterator());
+      }
+    }
+  }
+
+  @Override
+  public void rollback(TopicMetadata metadata, RollbackDetail rollbackDetail) throws IOException {
+    MessageCache<Entry> messageCache = cacheProvider.getMessageCache(metadata.getTopicId());
+    if (messageCache != null) {
+      // Rollback from the cache first so that we don't have to worry about invalid list pruning for the cache,
+      // assuming the rollback from cache shouldn't fail.
+      Entry startEntry = new CacheMessageTableEntry(metadata,
+                                                    rollbackDetail.getStartTimestamp(),
+                                                    (short) rollbackDetail.getStartSequenceId());
+      Entry endEntry = new CacheMessageTableEntry(metadata,
+                                                  rollbackDetail.getEndTimestamp(),
+                                                  (short) rollbackDetail.getEndSequenceId());
+      messageCache.updateEntries(startEntry, endEntry, new MessageCache.EntryUpdater<Entry>() {
+        @Override
+        public void updateEntry(Entry entry) {
+          if (!(entry instanceof CacheMessageTableEntry)) {
+            // This shouldn't happen
+            throw new IllegalStateException("Entries in MessageCache must be of type "
+                                              + CacheMessageTableEntry.class.getName()
+                                              + ", but got type " + entry.getClass().getName() + " instead.");
+          }
+          ((CacheMessageTableEntry) entry).rollback();
+        }
+      });
+    }
+
+    // Rollback from the table
+    messageTable.rollback(metadata, rollbackDetail);
+  }
+
+  @Override
+  public void close() throws IOException {
+    messageTable.close();
+  }
+
+  /**
+   * Returns {@code true} if the scanner created from the message cache contains all entries starting from the given
+   * start entry; otherwise return {@code false}.
+   */
+  private boolean cacheHasAllEntries(Entry startEntry, MessageCache.Scanner<Entry> scanner,
+                                     Comparator<MessageTable.Entry> comparator) {
+    Entry firstInCache = scanner.getFirstInCache();
+    return firstInCache != null && comparator.compare(firstInCache, startEntry) <= 0;
```

Comment:
Is it because if the firstInCache is smaller than the start entry, then the addAll would have added those entries to cache (since reduceInSize removes from the smallest to the larger entries in order?)